src/Tinfoil/Data/Hash.hs

```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+module Tinfoil.Data.Hash(
+    Hash(..)
+  , hexDigest
+) where
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.Text.Encoding as T
+
+import           GHC.Generics (Generic)
+
+import           P
+
+-- | Binary representation of a hash.
+newtype Hash =
+  Hash {
+    unHash :: ByteString
+  } deriving (Eq, Show, Generic)
+
+instance NFData Hash
```

Comment:
Hm actually, given our datatypes are `Strict` anyway, this may not actually be a problem even on 7.8.4. @jystic do you have an opinion on this?